File: src/daytrace/report.py
from collections import defaultdict
from datetime import date

from daytrace.models import (
    ActivityRecord,
    ActivityReport,
    ApplicationTotal,
    DayWindow,
    SourceKind,
)
from daytrace.transform import partition_window_seconds


SOURCE_ORDER = {
    SourceKind.WINDOW: 0,
    SourceKind.EDITOR: 1,
    SourceKind.BROWSER: 2,
}


def build_report(
    day: date,
    window: DayWindow,
    records: tuple[ActivityRecord, ...],
    project: str | None,
) -> ActivityReport:
    timeline = tuple(
        sorted(records, key=lambda item: (item.start, item.bucket_id, item.event_id))
    )
    allocation = partition_window_seconds(timeline)
    app_seconds: dict[str, float] = defaultdict(float)
    for record, seconds in allocation:
        app_seconds[record.app or "Unknown application"] += seconds
    applications = tuple(
        ApplicationTotal(app, seconds)
        for app, seconds in sorted(
            app_seconds.items(), key=lambda item: (-item[1], item[0].casefold(), item[0])
        )
    )
    sources = tuple(sorted({item.kind for item in timeline}, key=SOURCE_ORDER.__getitem__))
    return ActivityReport(
        day=day,
        timezone_name=window.timezone_name,
        project=project,
        active_seconds=sum(item.seconds for item in applications) if allocation else None,
        sources=sources,
        timeline=timeline,
        applications=applications,
    )

File: packages/daytrace-core/src/report.ts
import type {
  ActivityRecord,
  ActivityReport,
  ApplicationTotal,
  DayWindow,
  IsoDate,
  SourceKind,
} from "./models.js";
import { timestampMs } from "./time.js";
import { partitionWindowSeconds } from "./transform.js";

const SOURCE_ORDER: Readonly<Record<SourceKind, number>> = {
  "current-window": 0,
  editor: 1,
  browser: 2,
  afk: 3,
};

export function buildReport(
  day: IsoDate,
  window: DayWindow,
  records: readonly ActivityRecord[],
  project?: string,
): ActivityReport {
  const timeline = [...records].sort(
    (left, right) => timestampMs(left.start) - timestampMs(right.start)
      || left.bucketId.localeCompare(right.bucketId)
      || left.eventId.localeCompare(right.eventId),
  );
  const allocation = partitionWindowSeconds(timeline);
  const appSeconds = new Map<string, number>();
  for (const { record, seconds } of allocation) {
    const app = record.app ?? "Unknown application";
    appSeconds.set(app, (appSeconds.get(app) ?? 0) + seconds);
  }
  const applications: ApplicationTotal[] = [...appSeconds.entries()]
    .sort(([leftApp, leftSeconds], [rightApp, rightSeconds]) => rightSeconds - leftSeconds
      || leftApp.toLocaleLowerCase().localeCompare(rightApp.toLocaleLowerCase())
      || leftApp.localeCompare(rightApp))
    .map(([app, seconds]) => ({ app, seconds }));
  const sources = [...new Set(timeline.map((item) => item.kind))]
    .sort((left, right) => SOURCE_ORDER[left] - SOURCE_ORDER[right]);
  return {
    day,
    timezoneName: window.timezoneName,
    ...(project === undefined ? {} : { project }),
    ...(allocation.length === 0 ? {} : { activeSeconds: applications.reduce((total, item) => total + item.seconds, 0) }),
    sources,
    timeline,
    applications,
  };
}
